### ISSUE TITLE
Resource temporarily unavailable, 503 and socket-timeout

### ISSUE THREAD

**Issue by cboos** [GrahamDumpleton/mod_wsgi#181]
Hello Graham,

We faced an odd situation, in which our main site (trac.edgewall.org) would go offline for a period of 5 to 12 minutes, 5 or 6 times a day .

We are running the following software stack:
 - Debian GNU/Linux 8.6 (jessie)
 - apache2  2.4.10-10+deb8u7
 - libapache2-mod-wsgi 4.3.0-1 amd64

Relevant excerpt from the configuration:
 - `WSGIRestrictEmbedded On`
 - `WSGIApplicationGroup %{GLOBAL}`
 - `WSGIDaemonProcess trac processes=2 threads=1 maximum-requests=1000 display-name=%{GROUP}`
 - `Timeout 300` (!) (that seems to be the Ubuntu default)

### Analysis of the incident

The access.log shows during that period lots of 503 errors, and the error.log contained messages like the following:
```
[wsgi:error] [pid 12699:tid 140643816888064] (11)Resource temporarily unavailable: [client xxx.xxx.xxx.xxx:45614] mod_wsgi (pid=12699): Unable to connect to WSGI daemon process 'trac' on '/var/run/apache2/wsgi.251.33.2.sock'.
```
Meaning the wsgi daemon was stuck. Seeing that similar symptoms were reported 2 years ago on the mod_wsgi discussion list ([mod_wsgi returning 503 service unavailable](https://groups.google.com/d/topic/modwsgi/H7qPoqYNJdI/discussion)), I looked at the backtrace. It didn't show anything suspicious like libpq or mod_ssl, just plain mod_wsgi and apache2 code.

To have more precise debugging information, I rebuilt and installed the latest mod_wsgi, version 4.5.11 and got the following backtrace when the "blackout" happened again:
```
Thread 2 (Thread 0x7f0c6807a700 (LWP 5250)):
#0  0x00007f0c72914aed in poll () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007f0c72e20de3 in apr_poll () from /usr/lib/x86_64-linux-gnu/libapr-1.so.0
#2  0x00007f0c7373fb85 in ap_core_output_filter ()
#3  0x00007f0c6a942bea in wsgi_header_filter (f=0x7f0c5a1864b0, b=0x7f0c695227e8) at src/server/mod_wsgi.c:12081
#4  0x00007f0c6a95286e in Adapter_output (self=0x7f0c585e0030, data=0x1 <error: Cannot access memory at address 0x1>,
    data@entry=0x7f0c61680954 "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n<html xmlns=\"http://www.w3.org/1999/xhtml\">\n  \n  \n\n  \n\n\n  <head>\n    <title>\n      branche"..., length=300000, string_object=0x0, exception_when_aborted=4, exception_when_aborted@entry=1) at src/server/mod_wsgi.c:2387
#5  0x00007f0c6a952d12 in Adapter_write (self=0x7f0c585e0030, args=<optimized out>) at src/server/mod_wsgi.c:3369
#6  0x00007f0c6a51fe6c in PyEval_EvalFrameEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#7  0x00007f0c6a52027d in PyEval_EvalFrameEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#8  0x00007f0c6a593190 in PyEval_EvalCodeEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#9  0x00007f0c6a520171 in PyEval_EvalFrameEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#10 0x00007f0c6a593190 in PyEval_EvalCodeEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#11 0x00007f0c6a520171 in PyEval_EvalFrameEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#12 0x00007f0c6a593190 in PyEval_EvalCodeEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#13 0x00007f0c6a520171 in PyEval_EvalFrameEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#14 0x00007f0c6a593190 in PyEval_EvalCodeEx () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#15 0x00007f0c6a4e232c in ?? () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#16 0x00007f0c6a4eabe3 in PyObject_Call () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#17 0x00007f0c6a5926e7 in PyEval_CallObjectWithKeywords () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#18 0x00007f0c6a9535d9 in Adapter_run (object=<optimized out>, self=<optimized out>) at src/server/mod_wsgi.c:3121
#19 wsgi_execute_script (r=0x7f0c5a1850a0) at src/server/mod_wsgi.c:4068
#20 0x00007f0c6a956b2f in wsgi_hook_daemon_handler (c=<optimized out>) at src/server/mod_wsgi.c:12690
#21 wsgi_process_socket (daemon=<optimized out>, bucket_alloc=<optimized out>, sock=<optimized out>, p=<optimized out>) at src/server/mod_wsgi.c:8405
#22 wsgi_daemon_worker (thread=<optimized out>, p=<optimized out>) at src/server/mod_wsgi.c:8733
#23 wsgi_daemon_thread (thd=0x7f0c68078de0, data=0x1) at src/server/mod_wsgi.c:8820
#24 0x00007f0c72be80a4 in start_thread (arg=0x7f0c6807a700) at pthread_create.c:309
#25 0x00007f0c7291d62d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```
(similar trace for the other wsgi:trac process, which was also stuck, hence the blackout)

I was able to verify that this poll was blocking for 5 minutes, i.e. the Timeout value.
This means that the child worker wasn't responsive either during that period of time.

Looking closer at the access pattern during the day and in particular during the moments preceding the blackout, we noticed suspicious activity from a bot (YisouSpider) downloading .zip files from our repository /browser pages, blatantly ignoring our /robots.txt rules as it went.

It looked like that bot didn't bother to read the whole content of the .zip, or did it very slowly. In any case, that access pattern was reminiscent of slow HTTP read DoS.

To verify this, in another blackout period, I looked this time at where the *child worker* was blocked.
In many threads, it was blocked at the first recv in the wsgi_hook_handler, i.e. waiting for the responses for the requests queued on the wsgi daemon; one thread was waiting for establishing a connection to the wsgi daemon; 3 threads were blocked while *sending* back to the clients, like the following one:
```
Thread 15 (Thread 0x7f49f9bf5700 (LWP 23716)):
#0  0x00007f4a0f356aed in poll () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007f4a0f862de3 in apr_poll () from /usr/lib/x86_64-linux-gnu/libapr-1.so.0
#2  0x00007f4a10181b85 in ap_core_output_filter ()
#3  0x00007f4a07e1f438 in ?? () from /usr/lib/apache2/modules/mod_ssl.so
#4  0x00007f4a07e22bbd in ?? () from /usr/lib/apache2/modules/mod_ssl.so
#5  0x00007f4a07e1f77a in ?? () from /usr/lib/apache2/modules/mod_ssl.so
#6  0x00007f4a101711cc in ap_content_length_filter ()
#7  0x00007f4a0739321a in wsgi_transfer_response (buffer_size=<optimized out>, bb=<optimized out>, r=<optimized out>) at src/server/mod_wsgi.c:11082
#8  wsgi_execute_remote (r=0x6) at src/server/mod_wsgi.c:11930
#9  0x00007f4a07399296 in wsgi_hook_handler (r=0x7f49f0d0a0a0) at src/server/mod_wsgi.c:7131
#10 0x00007f4a101862a0 in ap_run_handler ()
#11 0x00007f4a101867e9 in ap_invoke_handler ()
#12 0x00007f4a1019c6c2 in ap_process_async_request ()
#13 0x00007f4a1019c860 in ap_process_request ()
#14 0x00007f4a10199162 in ?? ()
#15 0x00007f4a1018fb50 in ap_run_process_connection ()
#16 0x00007f4a08a7151b in ?? () from /usr/lib/apache2/modules/mod_mpm_worker.so
#17 0x00007f4a0f62a0a4 in start_thread (arg=0x7f49f9bf5700) at pthread_create.c:309
#18 0x00007f4a0f35f62d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

Looking at the code in both mod_wsgi.c and core_filters.c in Apache, it seems that there's no bug per se, the behavior we're seeing above is just the consequence of the fallback to a blocking write (`send_brigade_blocking`) when buckets have accumulated for too long.

### Workaround and proposal

This problematic access pattern and its consequences (5-10 min blackout periods several times a day) could be addressed by adapting the timeout. By using a short timeout of 5 seconds for the communication between the MPM worker and the wsgi daemon (`WSGIDaemonProcess  socket-timeout=5`), the problematic requests were aborted right away instead of after 5 minutes. This 5s "micro" downtime was not noticeable.

However, that also meant that legit requests that would take more than 5s to build up would now fail. We have quite some of them, it's not unusual that for some requests, between preparing the content for the template and rendering the template, it takes more than 10 or 20 seconds before the first byte is sent out by the wsgi daemon.

So here's my proposal: *why not use the socket-timeout only for the timeout used by the wsgi daemon when sending the response back to the child worker*? I don't think that the child worker thread expecting the response from the daemon would not be responsive for more than 5s (for example), unless it would be itself caught in a full long timeout due to a blocking write to a slow client, the precise situation we want to detect and abort as soon as possible. In the other direction, it seems sensible that the child worker gives the "usual" delay to the wsgi daemon to build up its response (the Apache `Timeout`).

I've successfully tested such an approach, and now trac.edgewall.org is again very responsive, had had no more blackout periods despite this stupid bot still trying 4-5 times per minute to download zip files it won't read, and we can still enjoy our slow template rendering time and access pages that take minutes to load...

Alternatively, if you don't want to change `socket-timeout` itself, then the next best thing might be to introduce another option (e.g. `socket-timeout-response`). The `timeout-*` space would start to get overcrowded, though, that's why I think getting the existing `socket-timeout` more "to the point" would be a better choice.

I would be very interested to hear your opinion on the issue, and whether you're interested in the patch.

**Comment by GrahamDumpleton** [comment]
Am surprised you see a (11)Resource temporarily unavailable error, but then I have never really understand the circumstances as to when that can come up.

The rest of the analysis seems reasonable and I concur that there is potential for an issue as you describe with a client that doesn't read the response, yet hangs, keeping the connection open.

A bit of history is necessary at this point.

In Apache 2.2 and older, when the mod_wsgi code which exists in the Apache child worker process and proxies the response gets data from the daemon process it would blindly output it back out to the HTTP client. The problem was that if the HTTP client was slow, or otherwise wasn't reading the response at a rate greater than what it was being read from the daemon process group, then the apache output bucket brigade would buffer up the data inside of the child worker process. The actions of the mod_wsgi code in reading more data from the daemon process would not be blocked. The consequence of this was that the size of the child worker processes would blow out where HTTP clients were slow. This is likely one of the big reasons why people saw excessive memory usage by Apache back in time. It was nothing to do with mod_wsgi per se, but due to the design of Apache and how it handled data being proxied through it.

Now Apache 2.4 fixed this issue by adding a measure of flow control in this arrangement.

What I did do is in mod_wsgi 4.4.7, I implemented my own version of flow control, overriding what Apache was using. This was firstly so that I eliminated the problem when mod_wsgi was used with Apache 2.2, but it also allowed me more control over how that flow control mechanism worked, allowing me to address what to me what was still a potential memory issue caused by a WSGI application returning data in lots of small blocks.

Either way, in Apache 2.4, or with mod_wsgi 4.4.7+ and Apache 2.2, that flow control means a blocking HTTP client would cause the child worker process to stop reading data returned from the daemon process and it would not now be buffered up in the child worker process, with the only buffering being a small amount (default of 64Kb with 4.4.7+, not sure about Apache 2.4 before that) in the child worker process and what would be buffered in the UNIX socket connection to the daemon process group.

Ultimately what it means is that the daemon process will block and can't offload the response and have the Apache child worker process worry about it all.

Now if using 4.4.7+ you could mediate the problem a little bit by increasing how much is buffered in memory by setting the 'response-buffer-size' option, but then you are back to potentially blowing out memory usage.  You could mediate against that a bit by setting Apache ThreadStackSize to lower value than default so that each request thread pool doesn't at least hold on to the memory and instead releases excess back to the process, but ultimately, if the response size is still greater than that value, it is going to block.

When things do block, the circuit breaker is going to be the Apache Timeout directive applied to the socket connection back to the client. This defaults to 300 seconds though which is ridiculous. People who know about setting up Apache would usually set this down at 60 seconds at least.

So when that timeout expires, the blocked write in the child worker process trying to flush data out to the client will fail, this will cause the proxying code to end the request and close the daemon process socket connection. The daemon process will then detect the close even though there is buffered data and so end the request in the daemon process.

Your solution has been to adjust the socket-timeout, which affects how long things will block on the socket connection to the daemon process from the child worker process. When reduced, this will at least free up the request thread in the daemon process, but the thread in the child worker process which is doing the proxying will not be, and it will be stuck until Timeout is reached. You could therefore still starve yourself of threads in the child worker processes.

I don't therefore think introducing a socket-timeout on the response is the answer, at least not with it applying to the daemon socket connection, as it will not unblock the thread in the child worker process talking to the client.

What would be best is a client-response-timeout, which overrides the Timeout value used by Apache in the child worker process. This would only apply to the specific case where using mod_wsgi daemon mode was doing the proxying and would be part of that special flow control code I have already added.

I just need to work out now how to apply the timeout in that case so let me look at and I'll see if I can work it out.

BTW, in mod_wsgi 4.5.12 there is a new restart-interval option which may be a better option than using maximum-requests=1000. The problem with maximum-requests was always that on a site that could get sudden bursts of traffic, you could start restarting processes at the worst time when have load. The restart-interval allows you to do restarts based on time instead of number of requests. Either way, you should look at using graceful-timeout option and using that at the same time to avoid interrupting running requests if running with multiple threads in same process.

**Comment by GrahamDumpleton** [comment]
The problem with overriding timeouts in this case has always been that for the HTTP client connection, the socket is technically private data that you aren't meant to deal with directly. They do seem to have added an API call for getting it in Apache 2.4 though, so if implement a client-response-timeout, then wouldn't be available in Apache 2.2, only Apache 2.4. That should be okay though.

**Comment by GrahamDumpleton** [comment]
Have implemented an option. Description from ``mod_wsgi-express`` help strings is:

```
    optparse.make_option('--response-buffer-size', type='int', default=0,
            metavar='NUMBER', help='Maximum amount of response content '
            'that will be allowed to be buffered in the Apache child '
            'worker process when proxying the response from a daemon '
            'process. Defaults to 0, indicating internal default of '
            '65536 bytes is used.'),
    optparse.make_option('--response-buffer-timeout', type='int', default=0,
            metavar='SECONDS', help='Maximum number of seconds allowed '
            'to pass before timing out on a write operation back to the '
            'HTTP client when the response buffer has filled and data is '
            'being forcibly flushed. Defaults to 0 seconds indicating that '
            'it will default to the value of the \'socket-timeout\' option.'),
```

For now naming it to parallel special ``response-buffer-size`` option since pertains to just that. It will take value of ``socket-timeout`` if not set explicitly. And if ``socket-timeout`` not set when doing manual configuration, uses whatever ``Timeout`` directive is set to.

Appears to work. I just need to refresh my memory about whether can just blindly set timeout and that only applies when socket in blocking mode and not non blocking mode.

**Comment by GrahamDumpleton** [comment]
Changed the option to ``response-socket-timeout``.

If there is a need be able to vary the socket timeout when sending content to the daemon process, or reading from it, will later add ``send-socket-timeout`` and ``receive-socket-timeout``. I can see a possible need for that for other reasons, but want to think about it.

For this particular issue, using ``response-socket-timeout`` to set timeout on socket used when sending response back to client from Apache child worker processes is still probably best best.

I have pushed up changes to develop branch of GitHub repo. Will ponder the changes a bit more and then make a release. Welcome to try by using:

* https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip

**Comment by GrahamDumpleton** [comment]
Released as 4.5.13.

**Comment by cboos** [comment]
It seems to work fine, thanks a lot!

With the "develop" version 4.5.13 installed, and the following configuration:
 - `WSGIDaemonProcess trac processes=6 threads=1 restart-interval=1500 response-socket-timeout=7  display-name=%{GROUP}`

I now get this in the error.log:
```
[Wed Jan 04 08:46:43.154956 2017] [wsgi:error] [pid 22894:tid 140237097363200] (70007)The timeout specified has expired: [client 42.120.161.61:33394] mod_wsgi (pid=22894): Failed to proxy response to client.
[Wed Jan 04 08:46:43.240083 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394] mod_wsgi (pid=22826): Exception occurred processing WSGI script '/var/www/edgewall.org/apache_trac.wsgi'.
[Wed Jan 04 08:46:43.240165 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394] Traceback (most recent call last):
[Wed Jan 04 08:46:43.240185 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 581, in dispatch_request
[Wed Jan 04 08:46:43.240217 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     return _dispatch_request(req, env, env_error)
[Wed Jan 04 08:46:43.240225 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 620, in _dispatch_request
[Wed Jan 04 08:46:43.240237 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     send_internal_error(env, req, sys.exc_info())
[Wed Jan 04 08:46:43.240245 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 751, in send_internal_error
[Wed Jan 04 08:46:43.240256 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     req.send_error(exc_info, status=500, env=env, data=data)
[Wed Jan 04 08:46:43.240289 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/api.py", line 732, in send_error
[Wed Jan 04 08:46:43.240302 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     exc_info)
[Wed Jan 04 08:46:43.240309 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 613, in _dispatch_request
[Wed Jan 04 08:46:43.240320 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     dispatcher.dispatch(req)
[Wed Jan 04 08:46:43.240327 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 253, in dispatch
[Wed Jan 04 08:46:43.240338 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     resp = chosen_handler.process_request(req)
[Wed Jan 04 08:46:43.240346 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/versioncontrol/web_ui/browser.py", line 402, in process_request
[Wed Jan 04 08:46:43.240358 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     self._render_zip(req, context, repos, node, rev)
[Wed Jan 04 08:46:43.240366 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/versioncontrol/web_ui/browser.py", line 659, in _render_zip
[Wed Jan 04 08:46:43.240377 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     render_zip(req, filename, repos, root_node, self._iter_nodes)
[Wed Jan 04 08:46:43.240384 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/versioncontrol/web_ui/util.py", line 231, in render_zip
[Wed Jan 04 08:46:43.240395 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     req.write(zip_str)
[Wed Jan 04 08:46:43.240402 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/api.py", line 833, in write
[Wed Jan 04 08:46:43.240413 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394]     self._write(''.join(buf))
[Wed Jan 04 08:46:43.240429 2017] [wsgi:error] [pid 22826:tid 140237293926144] [remote 42.120.161.61:33394] IOError: Apache/mod_wsgi failed to write response data: Broken pipe
[Wed Jan 04 08:46:43.565904 2017] [ssl:info] [pid 22894:tid 140237055403776] [client 163.172.66.153:29088] AH01964: Connection to child 17 established (server www.edgewall.org:443)
[Wed Jan 04 08:46:44.064754 2017] [wsgi:error] [pid 22895:tid 140237097363200] (70007)The timeout specified has expired: [client 42.156.138.61:52546] mod_wsgi (pid=22895): Failed to proxy response to client.
[Wed Jan 04 08:46:44.145791 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546] mod_wsgi (pid=22827): Exception occurred processing WSGI script '/var/www/edgewall.org/apache_trac.wsgi'.
[Wed Jan 04 08:46:44.145865 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546] Traceback (most recent call last):
[Wed Jan 04 08:46:44.145885 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 581, in dispatch_request
[Wed Jan 04 08:46:44.145917 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     return _dispatch_request(req, env, env_error)
[Wed Jan 04 08:46:44.145925 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 620, in _dispatch_request
[Wed Jan 04 08:46:44.145962 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     send_internal_error(env, req, sys.exc_info())
[Wed Jan 04 08:46:44.145971 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 751, in send_internal_error
[Wed Jan 04 08:46:44.145982 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     req.send_error(exc_info, status=500, env=env, data=data)
[Wed Jan 04 08:46:44.145990 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/api.py", line 732, in send_error
[Wed Jan 04 08:46:44.146001 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     exc_info)
[Wed Jan 04 08:46:44.146008 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 613, in _dispatch_request
[Wed Jan 04 08:46:44.146019 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     dispatcher.dispatch(req)
[Wed Jan 04 08:46:44.146026 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/main.py", line 253, in dispatch
[Wed Jan 04 08:46:44.146037 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     resp = chosen_handler.process_request(req)
[Wed Jan 04 08:46:44.146044 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/versioncontrol/web_ui/browser.py", line 402, in process_request
[Wed Jan 04 08:46:44.146057 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     self._render_zip(req, context, repos, node, rev)
[Wed Jan 04 08:46:44.146064 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/versioncontrol/web_ui/browser.py", line 659, in _render_zip
[Wed Jan 04 08:46:44.146075 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     render_zip(req, filename, repos, root_node, self._iter_nodes)
[Wed Jan 04 08:46:44.146083 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/versioncontrol/web_ui/util.py", line 231, in render_zip
[Wed Jan 04 08:46:44.146094 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     req.write(zip_str)
[Wed Jan 04 08:46:44.146101 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]   File "/usr/local/virtualenv/1.1dev/lib/python2.7/site-packages/trac/web/api.py", line 833, in write
[Wed Jan 04 08:46:44.146111 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546]     self._write(''.join(buf))
[Wed Jan 04 08:46:44.146128 2017] [wsgi:error] [pid 22827:tid 140237293926144] [remote 42.156.138.61:52546] IOError: Apache/mod_wsgi failed to write response data: Broken pipe
```
Which means there were two concurrent ".zip" requests which were successfully busted by the child worker due to the new timeout (`Failed to proxy response to client`).

Indeed, the access.log shows:
```
42.120.161.61 - - [04/Jan/2017:08:46:19 +0100] "GET /browser/tags/trac-0.9b1?format=zip&rev=4139 HTTP/1.1" 500 117130 "-" "YisouSpider"
163.172.66.86 - - [04/Jan/2017:08:46:43 +0100] "GET /babel/mirror/raw-annotate/5c55ca587c89/doc/epydoc.conf HTTP/1.1" 200 3860 "-" "Mozilla/5.0 (compatible; AhrefsBot/5.2; +http://ahrefs.com/robot/)"
68.180.228.51 - - [04/Jan/2017:08:46:43 +0100] "GET /ticket/11219?format=tab HTTP/1.1" 200 5231 "-" "Mozilla/5.0 (compatible; Yahoo! Slurp; http://help.yahoo.com/help/us/ysearch/slurp)"
163.172.66.153 - - [04/Jan/2017:08:46:43 +0100] "GET /genshi/genshi-test/annotate/3425c26d2c09/examples/turbogears/genshitest/static/images/favicon.ico HTTP/1.1" 200 5090 "-" "Mozilla/5.0 (compatible; AhrefsBot/5.2; +http://ahrefs.com/robot/)"
42.156.138.61 - - [04/Jan/2017:08:46:36 +0100] "GET /browser/tags/trac-0.9.4?format=zip&rev=6376 HTTP/1.1" 500 16073 "-" "YisouSpider"
```
Both `/browser?format=zip` requests from the YisouSpider failed with a 500, the first in (08:46:43 - 08:46:19 = 24) seconds, the second in (08:46:44 - 08:46:36 = 8) seconds, consistent with the `response-socket-timeout` of 7 seconds I used (the first client probably managed a few reads in time). A 408 failure in the log would probably be more appropriate, but I'm not sure if we can control that? We do receive a status of `APR_TIMEUP`, so I don't know why a 500 is logged and not a 408.

(And yes, in Trac we probably need to avoid sending an error in case of a Broken pipe IOError  ;-) )

My reasoning for focusing on the timeout used when the wsgi daemon writes the response to the worker child was that I considered the wsgi daemon to be the "valuable" resource and the child worker thread to be a "cheap" one. But that point of view is probably related to our specific needs or constraints. A single-threaded wsgi:trac process would typically take 150M of resident memory and be the one to burn the CPU, while an apache2 child worker process would only use 25M for 20-25 threads and use very few CPU time. So we easily had the potential to handle around 25 "stale" readers (~ 5 .zip requests per minute, Timeout of 300).

But in the bigger picture, you're probably right to also want to avoid waste at that level.
Which raises the question as to why Apache2 doesn't come with such a setting in the first place (TimeoutResponse directive?), as this seems to be generally useful beyond mod_wsgi! I found the [mod_reqtimeout](https://httpd.apache.org/docs/2.4/mod/mod_reqtimeout.html), but that's just about the request read side. Possibly [mod_qos](http://mod-qos.sourceforge.net/)? But that looks overkill compared to a simple response timeout, and even then, mod_qos doesn't seem to distinguish the request from the response timeout (single QS_Timeout).

By the way, could the `send-socket-timeout` or `receive-socket-timeout` you have in mind be used in place of my `socket-timeout-response` (send timeout from wsgi daemon to child worker)? It's not 100% clear what they would do and how they can be used, as send/receive depends on which side you're talking about (i.e. would `receive-socket-timeout` be used by both sides while they're receiving?), and there are many situations (i.e. would `receive-socket-timeout` also apply while the child worker is connecting?)

**Comment by GrahamDumpleton** [comment]
Stupid Google. Only just saw this as it went to spam.

I can record a different status code for the timeout case, but not sure that 408 is really the correct one. Status code of 504 is closer, but is really intended to relate to talking to a backend when acting as a gateway, not for the client.

The existing send-buffer-size relates to data going from Apache child worker to daemon process. The receive-buffer-size related to data going from daemon process to Apache child worker. I know the naming is a bit confusing but in in aligning the naming as best as could saw send-socket-timeout relating to the writing of data by Apache child worker to the daemon process, as well as the read timeout when daemon process reads request content from Apache child worker process. Only makes sense for the timeout to be same at both ends for the socket in one direction. The receive-socket-timeout would be for the response side of the socket.

So naming on that as it stands relates to the half of the socket connection as viewed from Apache child worker process.

You would be able to use receive-socket-timeout to achieve same end, but that isn't clearing it in the Apache child worker process at same time. Having the timeout on client side in Apache child worker process should propagate straight away to the connection to the daemon process, so don't see a downside to doing it that way.

**Comment by cboos** [comment]
Thanks for the additional details, and again thanks for the fix in the first place. It's working really great for us. From my side it's OK to close this ticket but in case you want me to test something else, I'll be pleased to do it.
